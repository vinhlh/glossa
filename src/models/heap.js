interface Item {
  id: string;
  score_days: float;
  score_random: int;
  history: Array<int>;
}

// Spaced repetition algorithm: SM-2
//
// Returns the number of days until seeing a problem again based on the
// history of answers x to the problem, where the meaning of x is:
// x == 0: Incorrect, Hardest
// x == 1: Incorrect, Hard
// x == 2: Incorrect, Medium
// x == 3: Correct, Medium
// x == 4: Correct, Easy
// x == 5: Correct, Easiest
// Based on https://gist.github.com/doctorpangloss/13ab29abd087dc1927475e560f876797
const getDaysTilNextTime = (
  history: Array<int>,
  a = 6.0,
  b = -0.8,
  c = 0.28,
  d = 0.02,
  assumed_score = 2.5,
  min_score = 1.3,
  theta = 0.2
): int => {
  if (!history.length) {
    return 1
  }

  // If got the last question incorrect, just return 1
  const last = history.length - 1
  if (history[last] < 3) {
    return 1
  }

  // Calculate the latest consecutive answer num_consecutively_correct
  let num_consecutively_correct = 0
  for (let i = last; i >= 0; i--) {
    if (history[i] >= 3) {
      num_consecutively_correct++
    } else {
      break
    }
  }

  // Sum up the history
  const historySum = history.reduce(
    (prev, val) => prev + (b + c * val + d * val * val),
    0
  )

  return (
    Math.round(
      a *
        Math.pow(
          Math.max(min_score, assumed_score + historySum),
          theta * num_consecutively_correct
        ) *
        1000
    ) / 1000
  )
}

const RANDOM_RANGE = 10

class MinHeap {
  items: Array<Item> = []

  pop(): Item {
    if (!this.items.length) {
      return null
    }

    const lastIndex = this.items.length - 1
    this._swap(this.items, 0, lastIndex)
    this._heapify(this.items, 0, lastIndex)

    const last = this.items[lastIndex]
    this.items.splice(lastIndex, 1)
    return last
  }

  push(item: Item) {
    this.items.push(item)

    let index = this.items.length - 1
    let parent

    while (true) {
      parent = this._parent(index)
      if (parent < 0) {
        return
      }

      if (this._less(this.items[parent], this.items[index])) {
        return
      }

      this._swap(this.items, index, parent)
      index = parent
    }
  }

  fromArray(items: Array<Item>): Array<Item> {
    let i = Math.floor(items.length / 2 - 1)

    while (i >= 0) {
      this._heapify(items, i, items.length)
      i -= 1
    }

    this.items = items
    return this.items
  }

  fromRawArray(rawItems: Array<Item>): Array<Item> {
    const items = rawItems.map(i => ({
      ...i,
      score_days: getDaysTilNextTime(i.history),
      score_random: this._random()
    }))

    return this.fromArray(items)
  }

  size(): int {
    return this.items.length
  }

  getItems(): Array<Item> {
    return this.items
  }

  _parent(index: int) {
    return Math.floor((index - 1) / 2)
  }

  _heapify(items: Array, i: int, max: int) {
    let index, left, right
    while (i < max) {
      index = i

      left = 2 * i + 1
      right = left + 1

      if (left < max && this._less(items[left], items[index])) {
        index = left
      }

      if (right < max && this._less(items[right], items[index])) {
        index = right
      }

      if (index === i) {
        return
      }

      this._swap(items, i, index)
      i = index
    }

    return items
  }

  _less(a: Item, b: Item) {
    if (a.score_days === b.score_days) {
      return a.score_random < b.score_random
    }

    return a.score_days < b.score_days
  }

  _swap(items: Array, a: int, b: int) {
    const t = items[a]
    items[a] = items[b]
    items[b] = t
  }

  _random() {
    return Math.random() * (RANDOM_RANGE - 0) + 0
  }
}

export { MinHeap, getDaysTilNextTime }
