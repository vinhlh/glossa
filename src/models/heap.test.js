import { MinHeap, getDaysTilNextTime } from './heap'

describe('MinHeap', () => {
  it('fromArray sorts empty array', () => {
    const h = new MinHeap()

    const items: Array<Item> = []

    const r = h.fromArray(items)
    expect(r).toMatchSnapshot()
  })

  it('fromArray sorts array with one item', () => {
    const h = new MinHeap()

    const items: Array<Item> = [
      {
        id: 'component',
        score_days: 5.0,
        score_random: 1
      }
    ]

    const r = h.fromArray(items)
    expect(r).toMatchSnapshot()
  })

  it('fromArray sorts items with multiple items have unique days properly', () => {
    const h = new MinHeap()

    const items: Array<Item> = [
      {
        id: 'component',
        score_days: 5.0,
        score_random: 1
      },
      {
        id: 'infrastructure',
        score_days: 4.0,
        score_random: 2
      },
      {
        id: 'hook',
        score_days: 3.0,
        score_random: 2
      },
      {
        id: 'walnut',
        score_days: 2.0,
        score_random: 2
      },
      {
        id: 'wonder',
        score_days: 1.0,
        score_random: 2
      }
    ]

    const r = h.fromArray(items)
    expect(r).toMatchSnapshot()
  })

  it('fromArray sorts items with multiple items have same days properly', () => {
    const h = new MinHeap()

    const items: Array<Item> = [
      {
        id: 'component',
        score_days: 2.0,
        score_random: 1
      },
      {
        id: 'infrastructure',
        score_days: 2.0,
        score_random: 3
      },
      {
        id: 'hook',
        score_days: 2.0,
        score_random: 2
      },
      {
        id: 'walnut',
        score_days: 2.0,
        score_random: 4
      },
      {
        id: 'wonder',
        score_days: 1.0,
        score_random: 5
      }
    ]

    const r = h.fromArray(items)
    expect(r).toMatchSnapshot()
  })

  it('fromArray sorts items with multiple items have same days, same random score properly', () => {
    const h = new MinHeap()

    const items: Array<Item> = [
      {
        id: 'component',
        score_days: 2.0,
        score_random: 3
      },
      {
        id: 'infrastructure',
        score_days: 2.0,
        score_random: 3
      },
      {
        id: 'hook',
        score_days: 2.0,
        score_random: 3
      },
      {
        id: 'walnut',
        score_days: 2.0,
        score_random: 3
      },
      {
        id: 'wonder',
        score_days: 1.0,
        score_random: 5
      }
    ]

    const r = h.fromArray(items)
    expect(r).toMatchSnapshot()
  })

  it('push then pop works correctly', () => {
    const h = new MinHeap()
    h.push({
      id: 'wonder',
      score_days: 1.0,
      score_random: 5
    })

    const popped = h.pop()
    expect(popped).toMatchSnapshot()
    expect(h.size()).toBe(0)
  })

  it('many pushes then pop works correctly', () => {
    const h = new MinHeap()
    h.push({
      id: 'wonder',
      score_days: 3.0,
      score_random: 5
    })
    h.push({
      id: 'walnut',
      score_days: 2.0,
      score_random: 3
    })
    h.push({
      id: 'infrastructure',
      score_days: 1.0,
      score_random: 3
    })
    expect(h.getItems()).toMatchSnapshot()

    const popped = h.pop()
    expect(popped).toMatchSnapshot()
    expect(h.getItems()).toMatchSnapshot()
  })

  it('so many pushes with many pop works correctly', () => {
    const h = new MinHeap()
    h.push({
      id: 'wonder',
      score_days: 3.0,
      score_random: 5
    })
    h.push({
      id: 'walnut',
      score_days: 2.0,
      score_random: 3
    })
    h.push({
      id: 'infrastructure',
      score_days: 8.0,
      score_random: 3
    })
    expect(h.getItems()).toMatchSnapshot()

    // pop one
    expect(h.pop()).toMatchSnapshot()
    expect(h.getItems()).toMatchSnapshot()

    // add 2 more
    h.push({ id: 'component', score_days: 4, score_random: 3.0 })
    h.push({ id: 'hook', score_days: 1, score_random: 3.0 })
    expect(h.getItems()).toMatchSnapshot()

    // 4 left
    expect(h.pop()).toMatchSnapshot()
    expect(h.getItems()).toMatchSnapshot()

    // 3 left
    expect(h.pop()).toMatchSnapshot()
    expect(h.getItems()).toMatchSnapshot()

    // 2 items left
    expect(h.pop()).toMatchSnapshot()
    expect(h.getItems()).toMatchSnapshot()

    // last element
    expect(h.pop()).toMatchSnapshot()
    expect(h.getItems()).toMatchSnapshot()

    // nothing in the heap
    expect(h.pop()).toMatchSnapshot()
    expect(h.getItems()).toMatchSnapshot()
  })
})

describe('getDaysTilNextTime', () => {
  test.each([
    [[], 1],
    [[0], 1],
    [[3], 7.329],
    [[0, 0, 1], 1],
    [[0, 0, 1, 2], 1],
    [[0, 1, 2, 3], 6.323],
    [[0, 1, 2, 3, 3], 7.019],
    [[0, 1, 2, 3, 2], 1],
    [[0, 1, 2, 3, 3, 5], 10.596],
    [[5, 5, 5, 5, 5, 5, 5], 154.95]
  ])(
    'getDaysTilNextTime(%s) returns correct days until next time',
    (input, expected) => {
      const got = getDaysTilNextTime(input)
      expect(got).toEqual(expected)
    }
  )
})
