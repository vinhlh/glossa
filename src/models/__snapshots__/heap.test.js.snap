// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MinHeap fromArray sorts array with one item 1`] = `
Array [
  Object {
    "id": "component",
    "score_days": 5,
    "score_random": 1,
  },
]
`;

exports[`MinHeap fromArray sorts empty array 1`] = `Array []`;

exports[`MinHeap fromArray sorts items with multiple items have same days properly 1`] = `
Array [
  Object {
    "id": "wonder",
    "score_days": 1,
    "score_random": 5,
  },
  Object {
    "id": "component",
    "score_days": 2,
    "score_random": 1,
  },
  Object {
    "id": "hook",
    "score_days": 2,
    "score_random": 2,
  },
  Object {
    "id": "walnut",
    "score_days": 2,
    "score_random": 4,
  },
  Object {
    "id": "infrastructure",
    "score_days": 2,
    "score_random": 3,
  },
]
`;

exports[`MinHeap fromArray sorts items with multiple items have same days, same random score properly 1`] = `
Array [
  Object {
    "id": "wonder",
    "score_days": 1,
    "score_random": 5,
  },
  Object {
    "id": "component",
    "score_days": 2,
    "score_random": 3,
  },
  Object {
    "id": "hook",
    "score_days": 2,
    "score_random": 3,
  },
  Object {
    "id": "walnut",
    "score_days": 2,
    "score_random": 3,
  },
  Object {
    "id": "infrastructure",
    "score_days": 2,
    "score_random": 3,
  },
]
`;

exports[`MinHeap fromArray sorts items with multiple items have unique days properly 1`] = `
Array [
  Object {
    "id": "wonder",
    "score_days": 1,
    "score_random": 2,
  },
  Object {
    "id": "walnut",
    "score_days": 2,
    "score_random": 2,
  },
  Object {
    "id": "hook",
    "score_days": 3,
    "score_random": 2,
  },
  Object {
    "id": "component",
    "score_days": 5,
    "score_random": 1,
  },
  Object {
    "id": "infrastructure",
    "score_days": 4,
    "score_random": 2,
  },
]
`;

exports[`MinHeap many pushes then pop works correctly 1`] = `
Array [
  Object {
    "id": "infrastructure",
    "score_days": 1,
    "score_random": 3,
  },
  Object {
    "id": "wonder",
    "score_days": 3,
    "score_random": 5,
  },
  Object {
    "id": "walnut",
    "score_days": 2,
    "score_random": 3,
  },
]
`;

exports[`MinHeap many pushes then pop works correctly 2`] = `
Object {
  "id": "infrastructure",
  "score_days": 1,
  "score_random": 3,
}
`;

exports[`MinHeap many pushes then pop works correctly 3`] = `
Array [
  Object {
    "id": "walnut",
    "score_days": 2,
    "score_random": 3,
  },
  Object {
    "id": "wonder",
    "score_days": 3,
    "score_random": 5,
  },
]
`;

exports[`MinHeap push then pop works correctly 1`] = `
Object {
  "id": "wonder",
  "score_days": 1,
  "score_random": 5,
}
`;

exports[`MinHeap so many pushes with many pop works correctly 1`] = `
Array [
  Object {
    "id": "walnut",
    "score_days": 2,
    "score_random": 3,
  },
  Object {
    "id": "wonder",
    "score_days": 3,
    "score_random": 5,
  },
  Object {
    "id": "infrastructure",
    "score_days": 8,
    "score_random": 3,
  },
]
`;

exports[`MinHeap so many pushes with many pop works correctly 2`] = `
Object {
  "id": "walnut",
  "score_days": 2,
  "score_random": 3,
}
`;

exports[`MinHeap so many pushes with many pop works correctly 3`] = `
Array [
  Object {
    "id": "wonder",
    "score_days": 3,
    "score_random": 5,
  },
  Object {
    "id": "infrastructure",
    "score_days": 8,
    "score_random": 3,
  },
]
`;

exports[`MinHeap so many pushes with many pop works correctly 4`] = `
Array [
  Object {
    "id": "hook",
    "score_days": 1,
    "score_random": 3,
  },
  Object {
    "id": "wonder",
    "score_days": 3,
    "score_random": 5,
  },
  Object {
    "id": "component",
    "score_days": 4,
    "score_random": 3,
  },
  Object {
    "id": "infrastructure",
    "score_days": 8,
    "score_random": 3,
  },
]
`;

exports[`MinHeap so many pushes with many pop works correctly 5`] = `
Object {
  "id": "hook",
  "score_days": 1,
  "score_random": 3,
}
`;

exports[`MinHeap so many pushes with many pop works correctly 6`] = `
Array [
  Object {
    "id": "wonder",
    "score_days": 3,
    "score_random": 5,
  },
  Object {
    "id": "infrastructure",
    "score_days": 8,
    "score_random": 3,
  },
  Object {
    "id": "component",
    "score_days": 4,
    "score_random": 3,
  },
]
`;

exports[`MinHeap so many pushes with many pop works correctly 7`] = `
Object {
  "id": "wonder",
  "score_days": 3,
  "score_random": 5,
}
`;

exports[`MinHeap so many pushes with many pop works correctly 8`] = `
Array [
  Object {
    "id": "component",
    "score_days": 4,
    "score_random": 3,
  },
  Object {
    "id": "infrastructure",
    "score_days": 8,
    "score_random": 3,
  },
]
`;

exports[`MinHeap so many pushes with many pop works correctly 9`] = `
Object {
  "id": "component",
  "score_days": 4,
  "score_random": 3,
}
`;

exports[`MinHeap so many pushes with many pop works correctly 10`] = `
Array [
  Object {
    "id": "infrastructure",
    "score_days": 8,
    "score_random": 3,
  },
]
`;

exports[`MinHeap so many pushes with many pop works correctly 11`] = `
Object {
  "id": "infrastructure",
  "score_days": 8,
  "score_random": 3,
}
`;

exports[`MinHeap so many pushes with many pop works correctly 12`] = `Array []`;

exports[`MinHeap so many pushes with many pop works correctly 13`] = `null`;

exports[`MinHeap so many pushes with many pop works correctly 14`] = `Array []`;
